docs/en/sql-reference/functions/date-time-functions.md

```diff
@@ -2553,8 +2553,8 @@ Using replacement fields, you can define a pattern for the resulting string. “
 | %i       | minute (00-59)                                          | 33         |
 | %I       | hour in 12h format (01-12)                              | 10         |
 | %j       | day of the year (001-366)                               | 002        |
-| %k       | hour in 24h format (00-23)                              | 22         |
-| %l       | hour in 12h format (01-12)                              | 09         |
+| %k       | hour in 24h format (0-23)                               | 2          |
```

Comment:
There are really two functions involved here: `parseDateTime()` and `formatDateTime()` and both functions "mirror" each other.

About `parseDateTime()`: `%l`/`%k` now accept single-digit hours. This is an incompatible change (as stated by the changelog entry) in a quite popular function. It would be better if there was a setting that restores the legacy. There are two reasons:
1. I know of situations in the past where incompatible output in functions broke existing use cases (queries, table/partition definitions). If this happens (and it does not really matter how unlikely we thing this is - there are always "weird misuses" of ClickHouse out there), a setting comes in handy.
2. The new behavior increases compatibility with other DBMS (good) but it also increases ambiguity in certain edge cases such as `SELECT parseDateTime('112', '%k%k');` - will the result have `2` or `12` as the hour component? Note that this can't happen with fixed-width formatters and we should give users a possibility to switch back.

About the name of the setting: As far as I see, `%c` (month) generally has the same problem: 2 digits in ClickHouse, 1-2 digits in MySQL's `FORMAT_DATE()` ([docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format)) function (after which ClickHouse's `formatDateTIme()` is modeled). `%c` is not addressed by this PR - if you like to, it would be very appreciated. So what about: `formatdatetime_lkc_is_space_padded` (default: `true`)? There could be a "Note 3" in this document, see l. 2578ff.

Then there is `formatDateTime()`: users expect that `parseDateTime(formatDateTime(...))` is idempotent, so similar change would happen there. `formatDateTime()` has [an optimization](https://github.com/ClickHouse/ClickHouse/blob/8c253b9e3e46c416623047d2df876eab0e2eb5ba/src/Functions/formatDateTime.cpp#L873) which will unfortunately no longer work for `%l`/`%k` (`%c`) but it is what it is. Ideally, the new setting is also respected in `formatDateTime()`.

Reply:
>1-2 digits in MySQL's FORMAT_DATE() ([docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format)) function (after which ClickHouse's formatDateTIme() is modeled)

Actually MySQL accept any variant

```sql
mysql> select version();
+-----------+
| version() |
+-----------+
| 8.2.0     |
+-----------+

mysql> SELECT STR_TO_DATE('1717:1', '%H');
+-----------------------------+
| STR_TO_DATE('1717:1', '%H') |
+-----------------------------+
| 17:00:00                    |
+-----------------------------+

mysql> SELECT STR_TO_DATE('7', '%H');
+------------------------+
| STR_TO_DATE('7', '%H') |
+------------------------+
| 07:00:00               |
+------------------------+
1 row in set (0.00 sec)

mysql> SELECT STR_TO_DATE('7', '%k');
+------------------------+
| STR_TO_DATE('7', '%k') |
+------------------------+
| 07:00:00               |
+------------------------+
1 row in set (0.00 sec)
```

Reply:
>So what about: formatdatetime_lkc_is_space_padded (default: true)

Looks too specific, I will go with `formatdatetime_parsedatetime_enable_format_with_leading_zeros`

Reply:
So:
- I've added support of month without leading zeros (`%c`)
- add setting to control behaviour of `parseDateTime` - `parsedatetime_enable_format_without_leading_zeros=1` (ON by default)
- adjust `formatDateTime` to this format changes as well, by only under `formatdatetime_enable_format_without_leading_zeros=1`, it is OFF by default (since this can break some user output, i.e. exports)
- adjusted settings changes
- updated the description (and of course commit message as well) to reflect this changes